### PR TITLE
CDAP-12325 do auth check before we do any drop

### PR DESCRIPTION
**Pull request by yaojiefeng**
JIRA: https://issues.cask.co/browse/CDAP-12325
Build: https://builds.cask.co/browse/CDAP-RUT1232-1

Artifact and dataset modules are already using this logic

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java [hunk 1]
```diff
@@ -536,17 +536,30 @@ public boolean apply(ProgramRuntimeService.RuntimeInfo runtimeInfo) {
...
+    removeApplication(appId, spec);
+  }
+
+  /**
+   * Remove application by the appId and appSpec, note that this method does not have any auth check
+   *
+   * @param applicationId the {@link ApplicationId} of the application to be removed
+   * @param appSpec the {@link ApplicationSpecification} of the application to be removed
+   */
+  private void removeApplication(ApplicationId applicationId, ApplicationSpecification appSpec) throws Exception {
```
Comment (poornachandra) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java]: Let's call this `removeAppInternal()` for readability

File: cdap-data-fabric/src/test/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetServiceAuthorizationTest.java [hunk 2]
```diff
@@ -168,20 +173,33 @@ public void execute() throws Exception {
...
+    dsFramework.deleteInstance(dsId);
+
+    grantAndAssertSuccess(dsId2, ALICE, EnumSet.of(Action.ADMIN));
+    // add add the instance again
+    dsFramework.addInstance(Table.class.getName(), dsId, DatasetProperties.EMPTY);
+    dsFramework.addInstance(Table.class.getName(), dsId1, DatasetProperties.EMPTY);
+    dsFramework.addInstance(Table.class.getName(), dsId2, DatasetProperties.EMPTY);
+
+    // should be successful since ALICE has ADMIN on all datasets
+    dsFramework.deleteAllInstances(NamespaceId.DEFAULT);
```
Comment (poornachandra) [inline cdap-data-fabric/src/test/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetServiceAuthorizationTest.java]: Before `deleteAllInstances()` let's list the instances and make sure there are three of them.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java [hunk 3]
```diff
@@ -490,12 +489,17 @@ public ApplicationWithPrograms deployApp(NamespaceId namespace, @Nullable String
...
    * @throws Exception
    */
   public void removeAll(final NamespaceId namespaceId) throws Exception {
-    List<ApplicationSpecification> allSpecs = new ArrayList<>(
-      store.getAllApplications(namespaceId));
+    List<ApplicationSpecification> allSpecs = new ArrayList<>(store.getAllApplications(namespaceId));
+    Map<ApplicationId, ApplicationSpecification> apps = new HashMap<>();
+    for (ApplicationSpecification appSpec : allSpecs) {
+      ApplicationId applicationId = namespaceId.app(appSpec.getName(), appSpec.getAppVersion());
+      authorizationEnforcer.enforce(applicationId, authenticationContext.getPrincipal(), Action.ADMIN);
```
Comment (poornachandra) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java:503]: Would be good to add a test for this

**Review by poornachandra — COMMENTED** [review]
@yaojiefeng Changes look good, just a couple of comments.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java (hunk 3 above)
Comment (yaojiefeng) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java:503]: This is tested here: https://github.com/caskdata/cdap/blob/release/4.3/cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java#L537

**Review by poornachandra — APPROVED** [review]

**Pull request merged**